pkg/compactor/blocks_cleaner.go

```diff
@@ -397,6 +405,39 @@ func (c *BlocksCleaner) deleteUserMarkedForDeletion(ctx context.Context, userID
 	return nil
 }
 
+// withRetries invokes the given function as many times as it takes according to
+// the backoff config. Each invocation will be given perCallTimeout to
+// complete. This is specifically designed to retry timeouts due to flaky
+// connectivity with the objstore backend.
+func withRetries(ctx context.Context, perCallTimeout time.Duration, f func(context.Context) error) error {
+	if perCallTimeout <= 0 {
+		return f(ctx)
+	}
+
+	var err error
+	b := backoff.New(ctx, backoff.Config{
+		MinBackoff: 20 * time.Millisecond,
+		MaxBackoff: 250 * time.Millisecond,
+		MaxRetries: 3,
+	})
+
+	for b.Ongoing() {
+		rctx, cancel := context.WithTimeout(ctx, perCallTimeout)
+		err = f(rctx)
+		cancel()
+		if err == nil || !shouldRetry(err) {
+			return err
+		}
+		b.Wait()
+	}
+
+	return fmt.Errorf("failed with retries: %w (last err: %w)", b.Err(), err)
+}
+
+func shouldRetry(err error) bool {
+	return errors.Is(err, context.DeadlineExceeded)
```

Comment:
the previous PR also used `var tempErr interface{ Temporary() bool }` as safe to retry on. Do you want to use the same here too?

Reply:
Sure, I'll add that back. GCS already has retries for errors that look like that, but minio does not.